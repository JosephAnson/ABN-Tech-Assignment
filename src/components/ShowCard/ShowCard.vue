<script setup lang="ts">
import Rating from "~/components/Rating/Rating.vue"
import type { ShowImage } from "~/types"

const props = defineProps<{
  id: number
  name: string
  summary: string
  rating?: {
    average?: number
  }
  image?: ShowImage
}>()
</script>

<template>
  <div class="bg-gray-800 rounded overflow-hidden h-full">
    <router-link :to="`/title/${props.id}`">
      <div class="image-container aspect-[42/59]">
        <img
          v-if="image"
          class="w-full object-cover h-full"
          :src="image"
          :alt="name"
        />
        <div v-if="!image" class="w-full bg-gray h-full flex" />
      </div>
      <div class="p-2">
        <Rating v-if="rating && rating.average">{{ rating.average }}</Rating>
        <Heading h4 bold class="truncate">
          {{ props.name }}
        </Heading>
        <div
          class="text-gray-400 text-xs overflow-hidden text-ellipsis text-clip h-12"
          v-html="props.summary"
        />
        <slot />
      </div>
    </router-link>
  </div>
</template>
