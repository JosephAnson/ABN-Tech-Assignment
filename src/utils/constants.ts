export const API_URL = "https://api.tvmaze.com/"
export const NOT_FOUND_IMAGE =
  "https://images.unsplash.com/photo-1461151304267-38535e780c79?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=300&q=100"
