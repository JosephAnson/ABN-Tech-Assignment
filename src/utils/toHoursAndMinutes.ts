export function toHoursAndMinutes(totalMinutes: number) {
  if (totalMinutes < 60) return `${totalMinutes}m`
  if (!(totalMinutes % 60)) return `${totalMinutes / 60}h`
  const hours = Math.floor(totalMinutes / 60)
  const minutes = totalMinutes % 60

  return `${hours}h ${minutes}m`
}
