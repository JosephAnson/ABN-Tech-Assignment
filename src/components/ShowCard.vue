<script setup lang="ts">
import Rating from "~/components/Rating.vue"
import { notFoundImage } from "~/utils/constants"

const props = defineProps<{
  id: number
  name: string
  summary?: string
  rating?: {
    average?: number
  }
  image?: string
}>()
</script>

<template>
  <div
    class="h-full rounded-xl overflow-hidden group bg-primary-500 hover:bg-primary-400 transition-250"
  >
    <router-link :to="`/title/${props.id}`" class="h-full">
      <div class="image-container aspect-[42/59] relative overflow-hidden">
        <img
          v-if="image"
          class="w-full h-full group-hover:scale-102 transition-250"
          :src="image"
          :alt="name"
        />
        <img
          v-if="!image"
          class="w-full h-full group-hover:scale-102 transition-250 object-cover"
          :src="notFoundImage"
          :alt="`No image for ${props.name}`"
        />
        <span
          class="overlay bg-gradient-to-t from-primary-900 via-transparent"
        ></span>
        <Rating
          v-if="rating && rating.average"
          class="position-absolute bottom-2 left-2"
          >{{ rating.average }}</Rating
        >
      </div>
      <div class="p-2 pointer-events-none">
        <Heading h4 bold class="truncate">
          {{ props.name }}
        </Heading>
        <div
          v-if="summary"
          class="text-gray-400 text-xs overflow-hidden text-ellipsis text-clip h-12"
          v-html="props.summary"
        />
        <slot />
      </div>
    </router-link>
  </div>
</template>
