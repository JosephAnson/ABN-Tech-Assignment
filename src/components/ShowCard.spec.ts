import { cleanup, render } from "@testing-library/vue"
import { afterEach, describe, expect, test } from "vitest"
import ShowCard from "./ShowCard.vue"
import { notFoundImage } from "~/utils/constants"

function createRenderShowCard(props: any) {
  return render(ShowCard, {
    props,
    global: {
      components: {
        RouterLink: {
          template: `<a><slot /></a>`,
        },
      },
    },
  })
}

describe("ShowCard", () => {
  const name = "Under the Dome"
  const rating = {
    average: 23,
  }
  const summary =
    "Under the Dome is the story of a small town that is suddenly and inexplicably sealed off from the rest of the world by an enormous transparent dome."
  const image = "http://thisisanimage.com/"

  afterEach(() => cleanup())

  test("should render show card", () => {
    const { html, getByText } = createRenderShowCard({
      id: 23,
      name,
    })
    expect(getByText(name)).toBeTruthy()
    expect(html()).toMatchSnapshot()
  })

  test("should render show card rating", () => {
    const { html, getByText } = createRenderShowCard({
      id: 23,
      name,
      rating,
    })
    expect(getByText(rating.average)).toBeTruthy()
    expect(html()).toMatchSnapshot()
  })

  test("should render show card summary", () => {
    const { html, getByText } = createRenderShowCard({
      id: 23,
      name,
      summary: `<p>${summary}</p>`,
    })
    expect(getByText(summary)).toBeTruthy()
    expect(html()).toMatchSnapshot()
  })

  test("should render show card image not found", () => {
    const { html, getByRole } = createRenderShowCard({
      id: 23,
      name,
    })
    expect(getByRole("img")).toBeTruthy()
    expect(getByRole<HTMLImageElement>("img").src).toBe(notFoundImage)
    expect(html()).toMatchSnapshot()
  })

  test("should render show card image", () => {
    const { html, getByRole } = createRenderShowCard({
      id: 23,
      name,
      image,
    })
    expect(getByRole("img")).toBeTruthy()
    expect(getByRole<HTMLImageElement>("img").src).toBe(image)
    expect(html()).toMatchSnapshot()
  })

  test("should render full card image", () => {
    const { html } = createRenderShowCard({
      id: 23,
      name,
      rating,
      image,
      summary,
    })
    expect(html()).toMatchSnapshot()
  })
})
