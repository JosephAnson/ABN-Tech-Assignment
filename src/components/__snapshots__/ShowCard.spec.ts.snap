// Vitest Snapshot v1

exports[`ShowCard > should render full card image 1`] = `
"<div class=\\"h-full rounded-xl overflow-hidden group bg-primary-500 hover:bg-primary-400 transition-250\\"><a to=\\"/title/23\\" class=\\"h-full\\">
    <div class=\\"image-container aspect-[42/59] relative overflow-hidden\\"><img class=\\"w-full h-full group-hover:scale-102 transition-250\\" src=\\"http://thisisanimage.com/\\" alt=\\"Under the Dome\\">
      <!--v-if--><span class=\\"overlay bg-gradient-to-t from-primary-900 via-transparent\\"></span>
      <div class=\\"flex items-center mb-1 last:mb-0 color-gray-300 position-absolute bottom-2 left-2\\"><span class=\\"i-carbon-star-filled color-yellow mr-1\\"></span><span>23</span></div>
    </div>
    <div class=\\"p-2 pointer-events-none\\">
      <p class=\\"heading font-secondary font-semibold md:text-l mb-2 font-bold truncate\\">Under the Dome</p>
      <div class=\\"text-gray-400 text-xs overflow-hidden text-ellipsis text-clip h-12\\">Under the Dome is the story of a small town that is suddenly and inexplicably sealed off from the rest of the world by an enormous transparent dome.</div>
    </div>
  </a></div>"
`;

exports[`ShowCard > should render show card 1`] = `
"<div class=\\"h-full rounded-xl overflow-hidden group bg-primary-500 hover:bg-primary-400 transition-250\\"><a to=\\"/title/23\\" class=\\"h-full\\">
    <div class=\\"image-container aspect-[42/59] relative overflow-hidden\\">
      <!--v-if--><img class=\\"w-full h-full group-hover:scale-102 transition-250 object-cover\\" src=\\"https://images.unsplash.com/photo-1461151304267-38535e780c79?ixlib=rb-4.0.3&amp;ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&amp;auto=format&amp;fit=crop&amp;w=300&amp;q=100\\" alt=\\"No image for Under the Dome\\"><span class=\\"overlay bg-gradient-to-t from-primary-900 via-transparent\\"></span>
      <!--v-if-->
    </div>
    <div class=\\"p-2 pointer-events-none\\">
      <p class=\\"heading font-secondary font-semibold md:text-l mb-2 font-bold truncate\\">Under the Dome</p>
      <!--v-if-->
    </div>
  </a></div>"
`;

exports[`ShowCard > should render show card image 1`] = `
"<div class=\\"h-full rounded-xl overflow-hidden group bg-primary-500 hover:bg-primary-400 transition-250\\"><a to=\\"/title/23\\" class=\\"h-full\\">
    <div class=\\"image-container aspect-[42/59] relative overflow-hidden\\"><img class=\\"w-full h-full group-hover:scale-102 transition-250\\" src=\\"http://thisisanimage.com/\\" alt=\\"Under the Dome\\">
      <!--v-if--><span class=\\"overlay bg-gradient-to-t from-primary-900 via-transparent\\"></span>
      <!--v-if-->
    </div>
    <div class=\\"p-2 pointer-events-none\\">
      <p class=\\"heading font-secondary font-semibold md:text-l mb-2 font-bold truncate\\">Under the Dome</p>
      <!--v-if-->
    </div>
  </a></div>"
`;

exports[`ShowCard > should render show card image not found 1`] = `
"<div class=\\"h-full rounded-xl overflow-hidden group bg-primary-500 hover:bg-primary-400 transition-250\\"><a to=\\"/title/23\\" class=\\"h-full\\">
    <div class=\\"image-container aspect-[42/59] relative overflow-hidden\\">
      <!--v-if--><img class=\\"w-full h-full group-hover:scale-102 transition-250 object-cover\\" src=\\"https://images.unsplash.com/photo-1461151304267-38535e780c79?ixlib=rb-4.0.3&amp;ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&amp;auto=format&amp;fit=crop&amp;w=300&amp;q=100\\" alt=\\"No image for Under the Dome\\"><span class=\\"overlay bg-gradient-to-t from-primary-900 via-transparent\\"></span>
      <!--v-if-->
    </div>
    <div class=\\"p-2 pointer-events-none\\">
      <p class=\\"heading font-secondary font-semibold md:text-l mb-2 font-bold truncate\\">Under the Dome</p>
      <!--v-if-->
    </div>
  </a></div>"
`;

exports[`ShowCard > should render show card rating 1`] = `
"<div class=\\"h-full rounded-xl overflow-hidden group bg-primary-500 hover:bg-primary-400 transition-250\\"><a to=\\"/title/23\\" class=\\"h-full\\">
    <div class=\\"image-container aspect-[42/59] relative overflow-hidden\\">
      <!--v-if--><img class=\\"w-full h-full group-hover:scale-102 transition-250 object-cover\\" src=\\"https://images.unsplash.com/photo-1461151304267-38535e780c79?ixlib=rb-4.0.3&amp;ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&amp;auto=format&amp;fit=crop&amp;w=300&amp;q=100\\" alt=\\"No image for Under the Dome\\"><span class=\\"overlay bg-gradient-to-t from-primary-900 via-transparent\\"></span>
      <div class=\\"flex items-center mb-1 last:mb-0 color-gray-300 position-absolute bottom-2 left-2\\"><span class=\\"i-carbon-star-filled color-yellow mr-1\\"></span><span>23</span></div>
    </div>
    <div class=\\"p-2 pointer-events-none\\">
      <p class=\\"heading font-secondary font-semibold md:text-l mb-2 font-bold truncate\\">Under the Dome</p>
      <!--v-if-->
    </div>
  </a></div>"
`;

exports[`ShowCard > should render show card summary 1`] = `
"<div class=\\"h-full rounded-xl overflow-hidden group bg-primary-500 hover:bg-primary-400 transition-250\\"><a to=\\"/title/23\\" class=\\"h-full\\">
    <div class=\\"image-container aspect-[42/59] relative overflow-hidden\\">
      <!--v-if--><img class=\\"w-full h-full group-hover:scale-102 transition-250 object-cover\\" src=\\"https://images.unsplash.com/photo-1461151304267-38535e780c79?ixlib=rb-4.0.3&amp;ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&amp;auto=format&amp;fit=crop&amp;w=300&amp;q=100\\" alt=\\"No image for Under the Dome\\"><span class=\\"overlay bg-gradient-to-t from-primary-900 via-transparent\\"></span>
      <!--v-if-->
    </div>
    <div class=\\"p-2 pointer-events-none\\">
      <p class=\\"heading font-secondary font-semibold md:text-l mb-2 font-bold truncate\\">Under the Dome</p>
      <div class=\\"text-gray-400 text-xs overflow-hidden text-ellipsis text-clip h-12\\">
        <p>Under the Dome is the story of a small town that is suddenly and inexplicably sealed off from the rest of the world by an enormous transparent dome.</p>
      </div>
    </div>
  </a></div>"
`;
